File: leetcode/easy/14_longest_common_prefix.py
# -----------------------------------------
# My Solution
#
# Time  Complexity: O(ns)
# Space Complexity: O(1)
# -----------------------------------------
# n := len(strs), s := min(string for string in strs)
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) == 0:
            return ""

        curr_pointer = 0
        while curr_pointer < len(strs[0]):
            curr_char = strs[0][curr_pointer]
            for index in range(1, len(strs)):
                if curr_pointer >= len(strs[index]) or strs[index][curr_pointer] != curr_char:
                    return strs[0][:curr_pointer]
            curr_pointer += 1
        return strs[0][:curr_pointer]

File: leetcode/easy/14_longest_common_prefix.cpp
// -----------------------------------------
// My Solution
//
// Time  Complexity: O(mn)
// Space Complexity: O(m)
// -----------------------------------------
// m := max string size in strs, n := strs.size()
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string longest_common_prefix = "";
        int str_cursor_position = 0;
        while (true) {
            char curr_char = '\0';
            for (int index = 0; index < strs.size(); ++index) {
                if (str_cursor_position >= strs[index].size())
                    return longest_common_prefix;

                if (curr_char == '\0') {
                    curr_char = strs[index][str_cursor_position];
                } else if (strs[index][str_cursor_position] != curr_char) {
                    return longest_common_prefix;
                }
            }

            longest_common_prefix += curr_char;
            str_cursor_position += 1;
        }
    }
};
